### PR TITLE
Make OnceCell<T> transparent to dropck

### PR DESCRIPTION
**Pull request by matklad**
See the failed build in

https://github.com/rust-lang/rust/pull/75555#issuecomment-675016718

for an example where we need this in real life

r? @ghost

### CODE REVIEW

**Comment by matklad** [comment]
For posterity, once_cell's PR: https://github.com/matklad/once_cell/pull/110/commits/d5295146e2fa80f4e7ad455fd97d21e3002ec885

**Comment by KodrAus** [comment]
Ah nice spotting!

@bors r+ rollup

**Comment by matklad** [comment]
cc https://github.com/rust-lang/rust/issues/75700

File: library/std/src/lazy.rs [hunk 1]
```diff
@@ -386,9 +386,10 @@ impl<T> SyncOnceCell<T> {
     }
 }
 
-impl<T> Drop for SyncOnceCell<T> {
+unsafe impl<#[may_dangle] T> Drop for SyncOnceCell<T> {
     fn drop(&mut self) {
-        // Safety: The cell is being dropped, so it can't be accessed again
+        // Safety: The cell is being dropped, so it can't be accessed again.
+        // We also don't touch the `T`, which validates our usage of #[may_dangle].
```
Comment (m-ou-se) [inline library/std/src/lazy.rs:392]: Looks like this isn't safe after all :(  See #76367

**Pull request merged**